### DISCUSSION TITLE
Received 1011 (internal error) - Deepgram did not receive audio data or a text message within the timeout window.

### DISCUSSION THREAD

**Discussion by devashish-gopalani-cognoai** [deepgram/community#540] [General help] [answered]
I have received an error while using deepgram's python SDK. This error has come after a considerate amount of time of establishing a socket connection with deepgram (close to more than a minute). Also I am sending a KeepAlive packet at an interval of 8 seconds. The error which I got was - 

```
Received 1011 (internal error) Deepgram did not receive audio data or a text message within the timeout window. See https://dpgr.am/net0001; then sent 1011 (internal error) Deepgram did not receive audio data or a text message within the timeout window. See https://dpgr.am/net0001 
```

In the documentation it has been written that - "If you’re receiving a 1011 code with a NET-0001 payload, Deepgram didn’t receive any audio from you before the system timed out."

I have the following questions:
1. What is the system time out duration?
2. Is this timeout applicable only when we don't send the first audio packet after the connection has been established?
3. Why is this issue coming if I am sending the KeepAlive packet at a regular duration?
4. How can this error be handled so that my functionality doesn't break?

I am using the public API of deepgram and the SDK version is 3.0.0

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by mounikawgsindia** [comment]
package com.example.deepgramdocwebsocket



import android.Manifest
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Bundle
import android.util.Log
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.deepgramdocwebsocket.MainActivity2.Companion.TAG
import kotlinx.coroutines.*
import okhttp3.*
import okio.ByteString
import okhttp3.WebSocket
import java.io.IOException

class MainActivity2 : AppCompatActivity() {

    companion object {
        val RECORD_AUDIO_PERMISSION_REQUEST_CODE = 123
        val TAG = "MainActivity"
        val SAMPLE_RATE = 44100
        private const val BYTES_PER_SAMPLE = 2
        val CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO
        val AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT
        private const val AUDIO_CHUNK_DURATION_MS = 20


        private val BUFFER_SIZE = calculateBufferSize(SAMPLE_RATE, AUDIO_CHUNK_DURATION_MS)

        private fun calculateBufferSize(sampleRate: Int, durationMs: Int): Int {
            val numSamples = (sampleRate * durationMs) / 1000
            return numSamples * CHANNEL_CONFIG * BYTES_PER_SAMPLE
        }


        //val BUFFER_SIZE = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL_CONFIG, AUDIO_FORMAT)
    }

    private lateinit var audioRecord: AudioRecord
    var isStreaming=false
    private lateinit var webSocket: WebSocket
    lateinit var startRecordingButton: Button
    lateinit var stopRecordingButton:Button
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        startRecordingButton = findViewById<Button>(R.id.startRecordingButton)
        stopRecordingButton = findViewById<Button>(R.id.stopRecordingButton)

        startRecordingButton.setOnClickListener {
            startRecording()
            connectWebSocket()
        }

        stopRecordingButton.setOnClickListener {
            stopRecording()
        }

        if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.RECORD_AUDIO
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                RECORD_AUDIO_PERMISSION_REQUEST_CODE
            )
        } else {
            //startRecording()
            //connectWebSocket()
        }
    }
    private fun stopRecording() {
        if (isStreaming) {
            audioRecord.stop()
            audioRecord.release()
            webSocket.close(1000, "Recording stopped")
        }

    }
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == RECORD_AUDIO_PERMISSION_REQUEST_CODE) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
              //  startRecording()
                //connectWebSocket()
            } else {
                // Handle permission denied
            }
        }
    }



    @android.annotation.SuppressLint("MissingPermission")
    private fun startRecording() {


        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            SAMPLE_RATE,
            CHANNEL_CONFIG,
            AUDIO_FORMAT,
            BUFFER_SIZE)

        if (audioRecord.state != AudioRecord.STATE_INITIALIZED) {
            Log.e("AudioRecord", "Failed to initialize AudioRecord")
            return
        }
        isStreaming = true
        audioRecord.startRecording()

        CoroutineScope(Dispatchers.IO).launch {
            sendAudioChunks()
        }
    }

    private fun sendAudioChunks() {
        val audioBuffer = ByteArray(BUFFER_SIZE)
        while (isStreaming) {
            val bytesRead = audioRecord.read(audioBuffer, 0, BUFFER_SIZE)
            if (bytesRead > 0) {
                sendAudio(audioBuffer.copyOfRange(0, bytesRead))
            }
        }
//        while (audioRecord.state == AudioRecord.STATE_INITIALIZED) {
//            val bytesRead = audioRecord.read(audioBuffer, 0, BUFFER_SIZE)
//            if (bytesRead > 0) {
//                sendAudio(audioBuffer.copyOfRange(0, bytesRead))
//            }
//        }
        if (audioRecord.state != AudioRecord.STATE_INITIALIZED) {
            Log.e(TAG, "Failed to initialize AudioRecord")

        }
    }

    private fun connectWebSocket() {
        val apiKey = "apikey"
        val client = OkHttpClient()
        val request = Request.Builder()
            .url("wss://api.deepgram.com/v1/listen")
            .addHeader("Authorization", "Token $apiKey")
            .build()

        val listener = DeepgramWebSocketListener()
        webSocket = client.newWebSocket(request, listener)
    }

    private fun sendAudio(audioBytes: ByteArray) {
        //Log.e(MainActivity2.TAG, "send byte array: ${audioBytes}")
        webSocket.send(ByteString.of(*audioBytes))
    }

    override fun onDestroy() {
        super.onDestroy()
        audioRecord.stop()
        audioRecord.release()
        webSocket.close(1000, "Activity destroyed")
    }
}

class DeepgramWebSocketListener : WebSocketListener() {
    override fun onOpen(webSocket: WebSocket, response: Response) {
        // Connection opened
        Log.e(MainActivity2.TAG, "WebSocket open: ${response}")
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        // Handle text messages (transcription results)
        Log.d(MainActivity2.TAG, "Transcription text: $text")
    }

    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
        // Handle binary messages if needed
        Log.d(MainActivity2.TAG, "Transcription: byte string $ByteString")
    }

    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        Log.d(MainActivity2.TAG, "Transcription: close1 $reason")
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        // Handle failure
        Log.e(MainActivity2.TAG, "WebSocket failure: ${t.message}")
        if (t is IOException && t.message?.contains("Broken pipe") == true) {
            // Handle "Broken pipe" error, e.g., attempt reconnection or notify user
            Log.e(TAG, "WebSocket connection failed: Broken pipe")
            // Attempt reconnection
            reconnectWebSocket()
        } else {
            // Handle other types of failures
            Log.e(TAG, "WebSocket failure: ${t.message}")
            // Notify user about the failure
            // For example, display a toast or show an error dialog
        }
    }
    private fun reconnectWebSocket() {
        val apiKey = "5f16450288c4af50a3ad457f313cfa3de8fbf342"
        val client = OkHttpClient()
        val request = Request.Builder()
            .url("wss://api.deepgram.com/v1/listen")
            .addHeader("Authorization", "Token $apiKey")
            .build()

        val newWebSocket = client.newWebSocket(request, this)
        // Optionally, update the WebSocket instance in the MainActivity
        // Replace the old WebSocket with the new one
        // (This depends on how your app manages WebSocket instances)
        // mainActivity.updateWebSocket(newWebSocket)
    }


}please tell me the answer

**Comment by davidvonthenen** [comment, marked as answer]
Looks like you opened an issue here:
https://github.com/deepgram/deepgram-python-sdk/issues/319

Closing this out. Please follow up on the issue.

**Comment by arpit-epsoft** [comment]
@devashish-gopalani-cognoai , I am also facing same issue. How did you resolved? @team-deepgram Can you help me how to resolve this?
